storage: don't violate raftMu < storeMu

tbg opened the pull request:
Before this change,

```
make stress TAGS=deadlock TESTS=TestCrashReportingPacket PKG=./pkg/util/log/
```

failed instantly (without the test code being at fault).

Fixes #18292.
Fixes #18294.
Fixes #18293.
Fixes #18295.
Fixes #18296.
Fixes #18297.
Fixes #18298.
Fixes #18299.
Fixes #18300.
Fixes #18301.
Fixes #18302.
Fixes #18303.
Fixes #18304.
Fixes #18305.
Fixes #18306.
Fixes #18307.
Fixes #18308.
Fixes #18309.
Fixes #18310.
Fixes #18311.
Fixes #18312.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18320)
<!-- Reviewable:end -->

petermattis commented on pkg/storage/store.go:
```diff
@@ -1118,9 +1117,19 @@ func (s *Store) Start(ctx context.Context, stopper *stop.Stopper) error {
+					s.mu.Unlock()
+					return false, err
+				}
+				s.mu.Unlock()
```
I'd write this as:

```
s.mu.Lock()
err := s.addReplicaInternalLocked(rep)
s.mu.Unlock()
if err != nil {
  return false, err
}
```

Do you know if there was something in the code that required holding `Store.mu` during the replica iteration?

tschottdorf replied:
Ah, good point. Done.

I checked the code and past diffs and I don't think anything else needs the lock.

tbg committed:
storage: don't violate raftMu < storeMu

Before this change,

```
make stress TAGS=deadlock TESTS=TestCrashReportingPacket PKG=./pkg/util/log/
```

failed instantly (without the test code being at fault).

Fixes #18292.
Fixes #18294.
Fixes #18293.
Fixes #18295.
Fixes #18296.
Fixes #18297.
Fixes #18298.
Fixes #18299.
Fixes #18300.
Fixes #18301.
Fixes #18302.
Fixes #18303.
Fixes #18304.
Fixes #18305.
Fixes #18306.
Fixes #18307.
Fixes #18308.
Fixes #18309.
Fixes #18310.
Fixes #18311.
Fixes #18312.

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18320#-:-KtRyldhkSxWFv1Tcmg0:bnfp4nl)*
<!-- Sent from Reviewable.io -->
